File: app/src/main/java/kapil/circularlayoutmanager/MainActivity.java
package kapil.circularlayoutmanager;

import android.os.Bundle;

import android.view.View;
import android.widget.TextView;
import android.widget.Toast;

import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.RecyclerView;

import com.google.android.material.floatingactionbutton.FloatingActionButton;
import com.kapil.circularlayoutmanager.CircularLayoutManager;
import com.kapil.circularlayoutmanager.OnItemClickListener;
import com.kapil.circularlayoutmanager.ScrollWheel;

import java.util.ArrayList;
import java.util.List;

/**
 * An example activity implementing a recycler view with a circular layout manager and scroll wheel.
 */

public class MainActivity extends AppCompatActivity implements View.OnClickListener {

    private RecyclerView recyclerView;
    private ScrollWheel scrollWheel;

    private FloatingActionButton addItemButton;
    private FloatingActionButton scrollWheelToggleButton;

    private List<Model> list;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        initViews();
        setViews();
    }

    private void initViews() {
        recyclerView = (RecyclerView) findViewById(R.id.recycler_view);
        scrollWheel = (ScrollWheel) findViewById(R.id.scroll_wheel);

        addItemButton = (FloatingActionButton) findViewById(R.id.add_item_button);
        scrollWheelToggleButton = (FloatingActionButton) findViewById(R.id.scroll_wheel_toggle_button);
    }

    private void setViews() {
        initializeList();
        recyclerView.setAdapter(new RecyclerViewAdapter(getApplicationContext(), list));
        recyclerView.addItemDecoration(new RecyclerItemDecoration());
        recyclerView.setLayoutManager(new CircularLayoutManager(getApplicationContext(), 200, -100));
        recyclerView.addOnItemTouchListener(new OnRecyclerItemClickListener(getApplicationContext(),
                new OnRecyclerItemClickListener.OnItemClickListener() {
            @Override
            public void OnItemClick(RecyclerView parent, int childIndex) {
                Toast.makeText(MainActivity.this, ((TextView) parent.getChildAt(childIndex)
                        .findViewById(R.id.event)).getText(), Toast.LENGTH_SHORT).show();
            }
        }));

        scrollWheel.setRecyclerView(recyclerView);
        scrollWheel.setScrollWheelEnabled(false);
        scrollWheel.setHighlightTouchAreaEnabled(false);
        //scrollWheel.setConsumeTouchOutsideTouchAreaEnabled(false);
        scrollWheel.setTouchAreaThickness(50);
        scrollWheel.setOnItemClickListener(new OnItemClickListener() {
            @Override
            public void onItemClick(ScrollWheel scrollWheel, int childIndex) {
                Toast.makeText(MainActivity.this, "OC " + ((TextView) scrollWheel.getRecyclerView()
                        .getChildAt(childIndex).findViewById(R.id.event)).getText(), Toast.LENGTH_SHORT).show();
            }

            @Override
            public void onItemLongClick(ScrollWheel scrollWheel, int childIndex) {
                Toast.makeText(MainActivity.this, "OLC " + ((TextView) scrollWheel.getRecyclerView()
                        .getChildAt(childIndex).findViewById(R.id.event)).getText(), Toast.LENGTH_SHORT).show();
            }
        });

        addItemButton.setOnClickListener(this);
        scrollWheelToggleButton.setOnClickListener(this);
    }

    private void initializeList() {
        list = new ArrayList<>();
        String event = "Event ", timing = "12:00am - 12:00pm";

        for (int i = 0; i < 10; i++) {
            Model model = new Model();
            model.setEvent(event + (i + 1));
            model.setTimings(timing);

            list.add(model);
        }
    }

    @Override
    public void onClick(View v) {
        switch (v.getId()) {

            case R.id.add_item_button:

                Model model = new Model();
                model.setEvent("Event " + (1 + list.size()));
                model.setTimings("12:00am - 12:00pm");

                list.add(model);

                recyclerView.getAdapter().notifyItemChanged(list.size() - 2);
                recyclerView.getAdapter().notifyItemInserted(list.size() - 1);

                break;

            case R.id.scroll_wheel_toggle_button:

                scrollWheel.setScrollWheelEnabled(!scrollWheel.isScrollWheelEnabled());
                scrollWheel.setHighlightTouchAreaEnabled(!scrollWheel.isHighlightTouchAreaEnabled());

                break;
        }
    }
}

File: app/src/main/java/kapil/circularlayoutmanager/MainActivity.kt
package kapil.circularlayoutmanager

import android.os.Bundle
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import com.kapil.circularlayoutmanager.INVALID_INDEX
import com.kapil.circularlayoutmanager.circularLayoutManager
import com.kapil.circularlayoutmanager.getChildAdapterPosition
import kotlinx.android.synthetic.main.activity_main.*

/**
 * An example activity implementing a recycler view with a circular layout manager and scroll wheel.
 *
 * In this example, the scroll wheel touch area partially lays over the recycler view.
 */
class MainActivity : AppCompatActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        initializeRecyclerView()
        initializeScrollWheel()

        addItemButton.setOnClickListener { addItemToList() }
        scrollWheelToggleButton.setOnClickListener { toggleScrollWheel() }
        smoothScrollTestBtn.setOnClickListener {
            if (positionInput.text.isNullOrBlank())
                showMessage("Enter an index")
            else
                recyclerView.smoothScrollToPosition(positionInput.text.toString().toInt())
        }
        scrollTestBtn.setOnClickListener {
            if (positionInput.text.isNullOrBlank())
                showMessage("Enter an index")
            else
                recyclerView.scrollToPosition(positionInput.text.toString().toInt())
        }
    }

    /**
     * The [com.kapil.circularlayoutmanager.CircularLayoutManager] is passed to the recycler view
     * in the XML layout file. The optional attributes along with constructor parameters are also
     * declared in the layout file.
     */
    private fun initializeRecyclerView() {
        recyclerView.adapter = RecyclerViewAdapter().apply {
            submitList(getInitialList())
            onItemClickListener = { showMessage(event) }
        }
        recyclerView.addItemDecoration(RecyclerItemDecoration())

//        recyclerView.layoutManager = CircularLayoutManager(
//            resources.getDimension(R.dimen.circular_list_radius),
//            resources.getDimension(R.dimen.circular_list_center_x)
//        ).apply {
//            scalingFactor = 1f
//            shouldIgnoreHeaderAndFooterMargins = true
//            shouldCenterIfProgrammaticallyScrolled = true
//            isAutoStabilizationEnabled = true
//        }
    }

    private fun initializeScrollWheel() {
        scrollWheel.isEnabled = false
        scrollWheel.isHighlightTouchAreaEnabled = false
//        scrollWheel.isHandleClicksEnabled = false
        scrollWheel.onItemClickListener = { x, y ->
            val index = recyclerView.getChildAdapterPosition(x, y)
            if (index != INVALID_INDEX) showMessage("onClick " + getList()[index].event)
        }
        scrollWheel.onItemLongClickListener = { x, y ->
            val index = recyclerView.getChildAdapterPosition(x, y)
            if (index != INVALID_INDEX) showMessage("onLongClick " + getList()[index].event)
        }
        scrollWheel.onScrollListener = { recyclerView.scrollBy(0, it.toInt()) }
        scrollWheel.onFlingListener = {
            // Since onTouchReleasedListener would be called before onFlingListener, we'll first
            // have to stop the ongoing stabilization before we could initiate the fling.
            recyclerView.stopScroll()
            recyclerView.fling(0, it.toInt())
        }
        scrollWheel.onTouchReleasedListener = {
            // Calling stabilize externally as scrolling and flinging is controlled by the
            // ScrollWheel. If RecyclerView.scrollBy method was not used, the external call to
            // stabilize is not necessary.
            recyclerView.circularLayoutManager!!.stabilize()
        }
    }

    private fun addItemToList() {
        (recyclerView.adapter as RecyclerViewAdapter).apply {
            submitList(currentList.toMutableList().apply {
                add(Model(size + 1, "Event ${size + 1}", "12:00am - 12:00pm"))
            }) { recyclerView.invalidateItemDecorations() }
        }
    }

    private fun toggleScrollWheel() {
        scrollWheel.isEnabled = !scrollWheel.isEnabled
        scrollWheel.isHighlightTouchAreaEnabled = !scrollWheel.isHighlightTouchAreaEnabled
    }

    private fun getInitialList() = (1..10).map { Model(it, "Event $it", "12:00am - 12:00pm") }

    private fun getList() = (recyclerView.adapter as RecyclerViewAdapter).currentList

    private fun showMessage(msg: String) =
        Toast.makeText(this@MainActivity, msg, Toast.LENGTH_SHORT).show()
}